File: src/main/java/com/codingforcookies/betterrecords/common/packets/PacketRadioPlay.java
package com.codingforcookies.betterrecords.common.packets;

import io.netty.buffer.ByteBuf;
import net.minecraft.entity.player.EntityPlayer;

import com.codingforcookies.betterrecords.client.sound.SoundHandler;

import net.minecraftforge.fml.common.network.ByteBufUtils;

public class PacketRadioPlay implements IPacket {
    int x, y, z, dimension;
    float playRadius;
    String localName, url;

    public PacketRadioPlay() { }

    public PacketRadioPlay(int x, int y, int z, float playRadius, int dimension, String localName, String url) {
        this.x = x;
        this.y = y;
        this.z = z;
        this.playRadius = playRadius;
        this.dimension = dimension;
        this.localName = localName;
        this.url = url;
    }

    public void readBytes(ByteBuf bytes) {
        String recieved = ByteBufUtils.readUTF8String(bytes);
        String[] str = recieved.split("\2477");
        x = Integer.parseInt(str[0]);
        y = Integer.parseInt(str[1]);
        z = Integer.parseInt(str[2]);
        playRadius = Float.parseFloat(str[3]);
        dimension = Integer.parseInt(str[4]);
        localName = str[5];
        url = str[6];
    }

    public void writeBytes(ByteBuf bytes) {
        ByteBufUtils.writeUTF8String(bytes, x + "\2477" + y + "\2477" + z + "\2477" + playRadius + "\2477" + dimension + "\2477" + localName + "\2477" + url);
    }

    public void executeClient(EntityPlayer player) {
        if(playRadius > 100000 || (float)Math.abs(Math.sqrt(Math.pow(player.posX - x, 2) + Math.pow(player.posY - y, 2) + Math.pow(player.posZ - z, 2))) < playRadius)
            SoundHandler.playSoundFromStream(x, y, z, dimension, playRadius, localName, url);
    }

    public void executeServer(EntityPlayer player) { }
}

File: src/main/kotlin/com/codingforcookies/betterrecords/network/PacketRadioPlay.kt
package com.codingforcookies.betterrecords.network

import com.codingforcookies.betterrecords.client.sound.SoundHandler
import io.netty.buffer.ByteBuf
import net.minecraft.client.Minecraft
import net.minecraft.util.math.BlockPos
import net.minecraftforge.fml.common.network.ByteBufUtils
import net.minecraftforge.fml.common.network.simpleimpl.IMessage
import net.minecraftforge.fml.common.network.simpleimpl.IMessageHandler
import net.minecraftforge.fml.common.network.simpleimpl.MessageContext

class PacketRadioPlay @JvmOverloads constructor(
        var pos: BlockPos = BlockPos(0, 0, 0),
        var dimension: Int = -1,
        var playRadius: Float = -1F,
        var local: String = "",
        var url: String = ""
) : IMessage {

    override fun toBytes(buf: ByteBuf) {
        buf.writeInt(pos.x)
        buf.writeInt(pos.y)
        buf.writeInt(pos.z)
        buf.writeInt(dimension)

        buf.writeFloat(playRadius)

        ByteBufUtils.writeUTF8String(buf, local)
        ByteBufUtils.writeUTF8String(buf, url)
    }

    override fun fromBytes(buf: ByteBuf) {
        pos = BlockPos(buf.readInt(), buf.readInt(), buf.readInt())
        dimension = buf.readInt()

        playRadius = buf.readFloat()

        local = ByteBufUtils.readUTF8String(buf)
        url = ByteBufUtils.readUTF8String(buf)
    }

    class Handler : IMessageHandler<PacketRadioPlay, IMessage> {

        override fun onMessage(message: PacketRadioPlay, ctx: MessageContext): IMessage? {
            val player = Minecraft.getMinecraft().player

            with (message) {
                if (playRadius > 100000 || Math.abs(Math.sqrt(Math.pow(player.posX - pos.x, 2.0) + Math.pow(player.posY - pos.y, 2.0) + Math.pow(player.posZ - pos.z, 2.0))).toFloat() < playRadius) {
                    SoundHandler.playSoundFromStream(pos, dimension, playRadius, local, url)
                }
            }

            return null
        }
    }
}